src/canvas/SelectableCanvas.ts

```diff
@@ -667,44 +672,22 @@ export class SelectableCanvas<
    * @return {Boolean}
    */
   isTargetTransparent(target: FabricObject, x: number, y: number): boolean {
-    // in case the target is the activeObject, we cannot execute this optimization
-    // because we need to draw controls too.
-    if (isFabricObjectCached(target) && target !== this._activeObject) {
-      // optimizatio: we can reuse the cache
-      const normalizedPointer = this._normalizePointer(target, new Point(x, y)),
-        targetRelativeX = Math.max(
-          target.cacheTranslationX + normalizedPointer.x * target.zoomX,
-          0
-        ),
-        targetRelativeY = Math.max(
-          target.cacheTranslationY + normalizedPointer.y * target.zoomY,
-          0
-        );
-
-      return isTransparent(
-        target._cacheContext,
-        Math.round(targetRelativeX),
-        Math.round(targetRelativeY),
-        this.targetFindTolerance
-      );
-    }
-
-    const ctx = this.contextCache,
-      originalColor = target.selectionBackgroundColor,
-      v = this.viewportTransform;
-
-    target.selectionBackgroundColor = '';
-
+    const ctx = this.pixelFindContext;
     this.clearContext(ctx);
-
     ctx.save();
-    ctx.transform(v[0], v[1], v[2], v[3], v[4], v[5]);
+    ctx.translate(-x + this.targetFindTolerance, -y + this.targetFindTolerance);
+    ctx.transform(...this.viewportTransform);
+    const selectionBgc = target.selectionBackgroundColor;
+    target.selectionBackgroundColor = '';
     target.render(ctx);
+    target.selectionBackgroundColor = selectionBgc;
     ctx.restore();
-
-    target.selectionBackgroundColor = originalColor;
-
-    return isTransparent(ctx, x, y, this.targetFindTolerance);
+    return isTransparent(
+      ctx,
+      this.targetFindTolerance,
+      this.targetFindTolerance,
+      this.targetFindTolerance
+    );
```

Comment:
This is the version of isTargetTransparent i ended up with, for the following reasons.
Still trying to figure out if retina should be included or not, my gut feeling is no, because is not part of the mouse interactions, is just a visual enhancement.
I left in the cache optimization because i was worried we could invalidate the cache with that render executed outside of context, didn't investigate deeply.
```js
  isTargetTransparent(target: FabricObject, x: number, y: number): boolean {
    const pixelCtx = this.pixelFindContext;
    const retina = this.getRetinaScaling();
    pixelCtx.resetTransform();
    this.clearContext(pixelCtx);
    const tolerance = this.targetFindTolerance;
    // normalized pointer is independendent from zoom and scale and retina.
    const normalizedPointer = this._normalizePointer(target, new Point(x, y));
    if (isFabricObjectCached(target) && !target.dirty) {
      // optimization: use the cache
      pixelCtx.translate(
        -target.cacheTranslationX / retina -
          (normalizedPointer.x * target.zoomX) / retina +
          tolerance,
        -target.cacheTranslationY / retina -
          (normalizedPointer.y * target.zoomY) / retina +
          tolerance
      );
      // the cache is retina scaled. we want to counter that.
      pixelCtx.scale(1 / retina, 1 / retina);
      // performance optimization:
      // we draw the hit area to the dedicated canvas instead of using `getImageData` on the target's cache canvas
      pixelCtx.drawImage(target._cacheCanvas, 0, 0);
    } else {
      pixelCtx.translate(-x + tolerance, -y + tolerance);
      pixelCtx.transform(...this.viewportTransform);
      target.render(pixelCtx);
    }
    return isTransparent(pixelCtx, tolerance, tolerance, tolerance);
  }
```

Reply:
this is just for reference, is missing the selection background color that you added back.

Reply:
My concern is testing a cached nested object. I don't think the drawing is correct in that situation. That is why I removed it completely. Also in case of a dirty cache the method will return a wrong value. So I thought it best to run and redraw cache if it needs

Reply:
i ll try to figure out if we can ever break the object cache by re-rendering a cached object.